In [1]:
!hpc

connect to node gpu02 of hpc provided by BDSRC


In [2]:
from sqlalchemy import create_engine


engine = create_engine('mysql+pyodbc://telecom')
%load_ext sql
%sql engine
%config SqlMagic.autopandas = True

Found pyproject.toml from '/Users/jacky/github/telecom'

In [3]:
%sql SHOW tables

Running query in 'mysql+pyodbc://telecom'

10 rows affected.

Tables_in_telecom
0                    cinfo_xy
1         sample_sight_201308
2          sample_user_201308
3  serv_acct_item_0838_201308
4  serv_acct_item_0838_201406
5     tb_asz_cdma_0838_201308
6     tb_asz_cdma_0838_201406
7                       tower
8                tower_center
9            tower_hot_201308

In [25]:
%%sql

DESC serv_acct_item_0838_201308

Running query in 'mysql+pyodbc://telecom'

20 rows affected.

Field         Type Null  Key Default Extra
0               SERV_ID   varchar(8)  YES  MUL    None      
1               ACC_NBR  varchar(64)  YES         None      
2               TRAFFIC          int  YES         None      
3              MONTH_NO   varchar(6)  YES         None      
4                CHARGE          int  YES         None      
5              DURATION          int  YES         None      
6   ACCT_ITEM_TYPE_CODE   varchar(6)  YES         None      
7            START_TIME  varchar(19)  YES         None      
8              END_TIME  varchar(19)  YES         None      
9           ETL_TYPE_ID   varchar(2)  YES         None      
10     CALLED_AREA_CODE   varchar(8)  YES         None      
11           CALLED_NBR  varchar(64)  YES         None      
12                  LAC   varchar(5)  YES  MUL    None      
13              CELL_ID   varchar(4)  YES  MUL    None      
14         RECV_TRAFFIC          int  YES         None      
15    CALLING_AREA_CODE   varchar(8)  YES         None      
16          CALLING_NBR  varchar(64)  YES         None      
17         SEND_TRAFFIC          int  YES         None      
18     BILLING_DURATION          int  YES         None      
19                CALLS   varchar(1)  YES            

# Calling Records

## Variables

- SERV_ID: user_id, can also be found in user information with the same column name
- ACC_NBR: cellphone number, can also be found in user infromation named as MSISDN
    - record the cellphone number of a user who is the client of the telecom company
- ACCT_ITEM_TYPE_CODE: 
- ETL_TYPE_ID: 
- CALLING_AREA_CODE, CALLED_AREA_CODE: determine the abstraction location of the calling and called phone call
- CALLING_NBR, CALLED_NBR: anonymous processed calling and called phone number
- LAC: local area code(most of them are outliers)
- CELL_ID: can be used for joing tower infromation on CELL16

- outlier
    - ACCT_ITEM_TYPE_CODE: -1
    - START_TIME, END_TIME: NOT LIKE '____/%/% %:__:__'
    - CALLING_AREA_CODE, CALLED_AREA_CODE: -1
    - CALLING_NBR, CALLED_NBR: NOT LIKE 'NotMobile_%'
    - LAC: -1, 0
    - CELL_ID: -1, 0

## Quickly look through the variables

In [5]:
%%sql

SELECT *
FROM serv_acct_item_0838_201308
LIMIT 5

Running query in 'mysql+pyodbc://telecom'

5 rows affected.

SERV_ID   ACC_NBR  TRAFFIC MONTH_NO  CHARGE  DURATION ACCT_ITEM_TYPE_CODE  \
0  a4zlm2z0  a1cnffy9        0   201308      12         0              020709   
1  a4zlm2z0  a1cnffy9        0   201308      12         1              020709   
2  a4zlm2z0  a1cnffy9        0   201308      12         0              020709   
3  a4zlm2z0  a1cnffy9        0   201308      12         0              020709   
4  a4zlm2z0  a1cnffy9        0   201308      12         0              020709   

         START_TIME          END_TIME ETL_TYPE_ID CALLED_AREA_CODE  \
0  2013/8/1 0:00:12  2013/8/1 0:00:12          23               -1   
1  2013/8/1 0:03:03  2013/8/1 0:03:04          23               -1   
2  2013/8/1 0:05:52  2013/8/1 0:05:52          23               -1   
3  2013/8/1 0:08:44  2013/8/1 0:08:44          23               -1   
4  2013/8/1 0:11:45  2013/8/1 0:11:45          23               -1   

     CALLED_NBR LAC CELL_ID  RECV_TRAFFIC CALLING_AREA_CODE   CALLING_NBR  \
0  NotMobile_-1  -1      -1             0                -1  NotMobile_-1   
1  NotMobile_-1  -1      -1             0                -1  NotMobile_-1   
2  NotMobile_-1  -1      -1             0                -1  NotMobile_-1   
3  NotMobile_-1  -1      -1             0                -1  NotMobile_-1   
4  NotMobile_-1  -1      -1             0                -1  NotMobile_-1   

   SEND_TRAFFIC  BILLING_DURATION CALLS  
0             0                 0     1  
1             0                 0     1  
2             0                 0     1  
3             0                 0     1  
4             0                 0     1

## Exception Detection

- DURATION doesn't have exception

In [4]:
%%sql

SELECT
    COUNT(*)
FROM serv_acct_item_0838_201308
WHERE DURATION < 0

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0         0

- exception of ACCT_ITEM_CODE is -1

In [11]:
%%sql

SELECT 
    ACCT_ITEM_TYPE_CODE,
    COUNT(*)
FROM serv_acct_item_0838_201308
GROUP BY  ACCT_ITEM_TYPE_CODE
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

48 rows affected.

ACCT_ITEM_TYPE_CODE  COUNT(*)
0               170211  55828623
1               170209  43815325
2               160201  25888559
3               160101  24548848
4               160204   3484611
5               160501   3392784
6               170205   2594883
7               170202   2450229
8               170208   2262250
9               160603   2089841
10              160604   1415474
11              170108    452489
12              120409    452409
13              020709    354989
14              170210    326767
15              020604    312595
16              020402    111271
17              100404     57270
18              020415     47031
19              080519     42251
20              120501     41031
21                  -1     27205
22              170204     19878
23              170207     16237
24              120101     10037
25              020410      9023
26              020305      8267
27              120405      6326
28              030105      6115
29              120202      5891
30              120209      5307
31              170212      3069
32              020601      2957
33              160402      1731
34              020309      1065
35              160302       969
36              020401       754
37              160203       439
38              170203       340
39              170201       310
40              170206       229
41              160202       216
42              160602       125
43              160601        67
44              020301        42
45              120201        11
46              021009         2
47              090101         1

- There are some invalid time foramt

In [14]:
%%sql

SELECT COUNT(*)
FROM serv_acct_item_0838_201308
WHERE 
    START_TIME NOT LIKE '____/%/% %:__:__' 
    OR
    END_TIME NOT LIKE '____/%/% %:__:__'

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0  44849841

- no exception in ETL_TYPE_ID

In [5]:
%%sql

SELECT
    ETL_TYPE_ID AS etl_type_id,
    COUNT(*)
FROM serv_acct_item_0838_201308
GROUP BY etl_type_id

Running query in 'mysql+pyodbc://telecom'

8 rows affected.

etl_type_id  COUNT(*)
0          21  60823665
1          22  60428727
2          23   4555058
3          24  43993499
4          31     50138
5          32    145182
6          33      6820
7          34     93054

- The CALLING_AREA_CODE and CALLED_AREA_CODES consist of 3 or 4 digit
- 0838 has the highest frequency

In [2]:
%%sql

SELECT
    CALLING_AREA_CODE,
    COUNT(*)
FROM serv_acct_item_0838_201308
WHERE CALLING_AREA_CODE != -1
GROUP BY CALLING_AREA_CODE
ORDER BY COUNG(*) DESC

Running query in 'mysql+pyodbc://telecom'

510 rows affected.

calling_area_code     count
0                0838  96205674
1                 028   3371312
2                0816    682887
3                 023    292693
4                 010    155611
..                ...       ...
505             00961         1
506             00973         1
507             08029         1
508             08071         1
509              0890         1

[510 rows x 2 columns]

In [5]:
%%sql

SELECT
    CALLING_AREA_CODE,
    COUNT(*)
FROM serv_acct_item_0838_201308
WHERE CALLING_AREA_CODE != -1
GROUP BY CALLING_AREA_CODE
ORDER BY COUNG(*) DESC
LIMIT 20

Running query in 'mysql+pyodbc://telecom'

20 rows affected.

CALLING_AREA_CODE     count
0               0838  96205674
1                028   3371312
2               0816    682887
3                023    292693
4                010    155611
5                021    155577
6               0817    127257
7               0755    110263
8               0825    106029
9               0833     99333
10               020     92556
11               029     87438
12              028M     87384
13              0839     84135
14              028Z     81907
15              0834     76948
16             0832a     76653
17              0830     74566
18              0871     74498
19              0813     71219

In [9]:
%%sql

SELECT
    CALLING_AREA_CODE,
    COUNT(*)
FROM serv_acct_item_0838_201308
WHERE CALLING_AREA_CODE != -1 AND LENGTH(CALLING_AREA_CODE) < 3
GROUP BY CALLING_AREA_CODE
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

Empty DataFrame
Columns: []
Index: []

In [11]:
%%sql

SELECT
    CALLED_AREA_CODE,
    COUNT(*) AS count
FROM serv_acct_item_0838_201308
WHERE CALLED_AREA_CODE != -1
GROUP BY CALLED_AREA_CODE
ORDER BY count DESC
LIMIT 20

Running query in 'mysql+pyodbc://telecom'

20 rows affected.

CALLED_AREA_CODE     count
0              0838  93860363
1               028   3794338
2              0816    761832
3               023    320694
4              0817    147813
5               010    134660
6              0825    129196
7               021    119936
8              0833    107266
9              0755    101168
10             0839     95790
11             0813     85704
12             0830     84671
13             0834     82676
14             0831     80925
15             0818     74431
16              029     69990
17              020     69563
18             0835     69040
19             0871     60194

In [14]:
%%sql

SELECT
    CALLED_AREA_CODE,
    COUNT(*) AS count
FROM serv_acct_item_0838_201308
WHERE CALLED_AREA_CODE != -1 AND LENGTH(CALLED_AREA_CODE) < 3
GROUP BY CALLED_AREA_CODE
ORDER BY count DESC

Running query in 'mysql+pyodbc://telecom'

Empty DataFrame
Columns: []
Index: []

- the exception are -1, 0 fro LAC
- most of the observations's LAC is exception

In [6]:
%%sql

SELECT
    LAC AS lac,
    COUNT(*)
FROM serv_acct_item_0838_201308
GROUP BY lac

Running query in 'mysql+pyodbc://telecom'

115 rows affected.

lac   COUNT(*)
0       -1  109329163
1        0    2336979
2    00000      46453
3    00001        214
4    00003         79
..     ...        ...
110  0264A          1
111  02A77          1
112  03133          1
113  09000          1
114  09008          1

[115 rows x 2 columns]

- most of the observation's CELL_ID is exception, namely the value is -1 or 0

In [15]:
%%sql

SELECT
    CELL_ID,
    COUNT(*)
FROM serv_acct_item_0838_201308
GROUP BY CELL_ID

Running query in 'mysql+pyodbc://telecom'

57001 rows affected.

CELL_ID  COUNT(*)
0          -1  48985582
1           0   2336698
2        0000     47155
3        0001      2184
4        0002      1151
...       ...       ...
56996    FFF2        32
56997    FFF3         2
56998    FFF7       249
56999    FFF9         2
57000    FFFB         6

[57001 rows x 2 columns]

##  try to figure out the period of the records

In [3]:
%%sql

SELECT DISTINCT
    MONTH_NO
FROM serv_acct_item_0838_201308

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

MONTH_NO
0   201308

In [6]:
%%sql 

SELECT
    date, 
    COUNT(*) AS count
FROM (
    SELECT SUBSTRING_INDEX(START_TIME,' ',1) AS date
    FROM serv_acct_item_0838_201308
) AS subquery
GROUP BY date

Running query in 'mysql+pyodbc://telecom'

31 rows affected.

date    count
0    2013/8/1  6179713
1   2013/8/10  5070191
2   2013/8/11  5010373
3   2013/8/12  5501876
4   2013/8/13  5760491
5   2013/8/14  5413238
6   2013/8/15  5415015
7   2013/8/16  5758206
8   2013/8/17  5054504
9   2013/8/18  4982765
10  2013/8/19  5370918
11   2013/8/2  5904688
12  2013/8/20  5688655
13  2013/8/21  5371282
14  2013/8/22  5483538
15  2013/8/23  5696578
16  2013/8/24  5091882
17  2013/8/25  4939744
18  2013/8/26  5693674
19  2013/8/27  5472572
20  2013/8/28  5455958
21  2013/8/29  5560549
22   2013/8/3  5631779
23  2013/8/30  5569868
24  2013/8/31  5393442
25   2013/8/4  5211282
26   2013/8/5  6015355
27   2013/8/6  5456617
28   2013/8/7  5667358
29   2013/8/8  5634673
30   2013/8/9  5639359

In [7]:
%%sql 

SELECT
    date, 
    COUNT(*) AS count
FROM (
    SELECT SUBSTRING_INDEX(START_TIME,' ',1) AS date
    FROM serv_acct_item_0838_201406
) AS subquery
GROUP BY date

Running query in 'mysql+pyodbc://telecom'

33 rows affected.

date    count
0   2014/5/29        3
1   2014/5/30        1
2   2014/5/31     1396
3    2014/6/1  6079751
4   2014/6/10  5354899
5   2014/6/11  5387891
6   2014/6/12  5223904
7   2014/6/13  5434661
8   2014/6/14  5064052
9   2014/6/15  5159278
10  2014/6/16  5635044
11  2014/6/17  5292012
12  2014/6/18  5207614
13  2014/6/19  5278601
14   2014/6/2  5675091
15  2014/6/20  5554002
16  2014/6/21  5114460
17  2014/6/22  4940918
18  2014/6/23  5605927
19  2014/6/24  5188945
20  2014/6/25  5540208
21  2014/6/26  6129365
22  2014/6/27  5593152
23  2014/6/28  5073869
24  2014/6/29  4958139
25   2014/6/3  5439193
26  2014/6/30  5416122
27   2014/6/4  5243489
28   2014/6/5  5346158
29   2014/6/6  5492140
30   2014/6/7  4884044
31   2014/6/8  4769464
32   2014/6/9  5150233

## Inspect the meaning of some column

In [3]:
%config SqlMagic.named_parameters=True

In [15]:
%%sql

CREATE TABLE clean_cdr(
    SELECT
        B.CI_BRANCH                                                                   AS branch,
        B.CI_DISTRICT                                                                 AS district,
        A.SERV_ID                                                                     AS id,
        A.ACC_NBR                                                                     AS client_nbr, 
        CONVERT(SUBSTRING_INDEX(SUBSTRING_INDEX(A.END_TIME,' ',-1),':',1),UNSIGNED)   AS time,
        DAYOFWEEK(SUBSTRING_INDEX(A.END_TIME,' ',1))                                  AS day_of_week,
        A.CALLING_NBR                                                                 AS calling_nbr,
        A.CALLING_AREA_CODE                                                           AS calling_area_code,
        A.DURATION                                                                    AS duration,
        A.CALLED_AREA_CODE                                                            AS called_area_code,
        A.CALLED_NBR                                                                  AS called_nbr, 
        A.CHARGE                                                                      AS charge, 
        A.ACCT_ITEM_TYPE_CODE                                                         AS acct_item_type_code, 
        A.ETL_TYPE_ID                                                                 AS etl_type_id,
        A.CELL_ID                                                                     AS cell_id
    FROM serv_acct_item_0838_201308 AS A 
    LEFT JOIN tb_asz_cdma_0838_201308 B ON A.SERV_ID=B.SERV_ID AND A.ACC_NBR=B.MSISDN
    WHERE 
        A.ACCT_ITEM_TYPE_CODE != -1                                                    AND
        (A.START_TIME LIKE "____/%/% %:__:__" AND A.END_TIME LIKE "____/%/% %:__:__")  AND
        (A.CALLING_NBR NOT LIKE 'NotMobile_%' AND A.CALLED_NBR NOT LIKE 'NotMobile_%') AND
        (A.CALLING_AREA_COdE != '-1' AND A.CALLED_AREA_CODE != '-1')                   AND
        (A.LAC != '-1' AND A.LAC != '0')                                               AND
        (A.CELL_ID != '-1' AND A.CELL_ID != '0')
)

Running query in 'mysql+pyodbc://telecom'

53121025 rows affected.

Empty DataFrame
Columns: []
Index: []

In [16]:
%sql desc clean_cdr

Running query in 'mysql+pyodbc://telecom'

15 rows affected.

Field             Type Null Key Default Extra
0                branch     varchar(255)  YES        None      
1              district     varchar(255)  YES        None      
2                    id       varchar(8)  YES        None      
3            client_nbr      varchar(64)  YES        None      
4                  time  bigint unsigned  YES        None      
5           day_of_week              int  YES        None      
6           calling_nbr      varchar(64)  YES        None      
7     calling_area_code       varchar(8)  YES        None      
8              duration              int  YES        None      
9      called_area_code       varchar(8)  YES        None      
10           called_nbr      varchar(64)  YES        None      
11               charge              int  YES        None      
12  acct_item_type_code       varchar(6)  YES        None      
13          etl_type_id       varchar(2)  YES        None      
14              cell_id       varchar(4)  YES        None

In [10]:
%sql SELECT * FROM clean_cdr LIMIT 60

Running query in 'mysql+pyodbc://telecom'

60 rows affected.

branch district        id client_nbr  time  day_of_week calling_nbr  \
0   德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr     9            5    eh2djdgr   
1   德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr     9            5    akmslxqx   
2   德阳市市辖区     德阳现业  9vei43io   itp9jhv5     9            5    itp9jhv5   
3   德阳市市辖区     德阳现业  9vei43io   itp9jhv5     9            5    itp9jhv5   
4   德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr     9            5    4dbnlczn   
5   德阳市市辖区     德阳现业  9vei43io   itp9jhv5    10            5    2fsek3dy   
6   德阳市市辖区     德阳现业  9vei43io   itp9jhv5    10            5    itp9jhv5   
7   德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    11            5    7wps2c9o   
8   德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    10            5    7wps2c9o   
9   德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    11            5    d59t2d4b   
10  德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    11            5    g6vxjd2w   
11  德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    11            5    g6vxjd2w   
12  德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    11            5    g6vxjd2w   
13  德阳市市辖区     德阳现业  9vei43io   itp9jhv5    11            5    27h8fuzu   
14  德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    11            5    g6vxjd2w   
15  德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    11            5    g6vxjd2w   
16  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    11            5    acy4izhx   
17  德阳市市辖区     德阳现业  9vei43io   itp9jhv5    11            5    itp9jhv5   
18  德阳市市辖区     德阳现业  9vei43io   itp9jhv5    11            5    2qu42rry   
19  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    11            5    eh2djdgr   
20  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    11            5    eh2djdgr   
21  德阳市市辖区     德阳现业  9vei43io   itp9jhv5    12            5    27h8fuzu   
22  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    12            5    eh2djdgr   
23  德阳市市辖区       罗江  a2yq2tls   nh51jdgv    12            5    nh51jdgv   
24  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    12            5    eh2djdgr   
25  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    12            5    b9s1bsbe   
26  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    13            5    kn4854qi   
27  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    13            5    kn4854qi   
28  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    13            5    aobsa8j1   
29  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    13            5    eh2djdgr   
30  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    14            5    b4512k7u   
31  德阳市市辖区     德阳现业  9vei43io   itp9jhv5    14            5    2fsek3dy   
32  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    14            5    eh2djdgr   
33  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    14            5    mwrc99yr   
34  德阳市市辖区     德阳现业  90m62trl   g6vxjd2w    15            5    e0a34eqb   
35  德阳市市辖区     德阳现业  9vei43io   itp9jhv5    16            5    27zsk3dy   
36  德阳市市辖区     德阳现业  9vei43io   itp9jhv5    17            5    itp9jhv5   
37  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    17            5    eh2djdgr   
38  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    17            5    m1fc4wpn   
39  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    17            5    m1fc4wpn   
40  德阳市市辖区     德阳现业  8kvi2lt9   eh2djdgr    17            5    eh2djdgr   
41  德阳市市辖区     德阳现业  9vei43io   itp9jhv5    21            5    itp9jhv5   
42  德阳市市辖区       罗江  a2yq2tls   nh51jdgv    21            5    3xvs5zaa   
43  德阳市市辖区     德阳现业  aibi2trk   2ozqknoi     7            5    2ozqknoi   
44  德阳市市辖区     德阳现业  8sla2tga   czd2kn7z     8            5    2z5s5clm   
45  德阳市市辖区     德阳现业  8sla2tga   czd2kn7z     9            5    9e9k3uih   
46  德阳市市辖区     德阳现业  8sla2tga   czd2kn7z     9            5    9e9k3uih   
47  德阳市市辖区     德阳现业  aibi2trk   2ozqknoi    10            5    5nlr2i4r   
48  德阳市市辖区     德阳现业  8kvi2tu5   desmknon    10            5    desmknon   
49  德阳市市辖区     德阳现业  9nge2tot   f4iuknlw    14            5    f4iuknlw   
50  德阳市市辖区     德阳现业  8sla2tga   czd2kn7z    14            5    amh499t1   
51  德阳市市辖区     德阳现业  9nge2tot   f4iuknlw    15            5    f4iuknlw   
52  德阳市市辖区     德阳现业  8kvi2tu4

- the isn't case that the client number is neither calling number nor called number(occurs in raw data)

In [19]:
%sql SELECT COUNT(*) FROM clean_cdr WHERE client_nbr != calling_nbr AND client_nbr != called_nbr

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0         0

In [7]:
%%sql

SELECT
    branch,
    COUNT(*)
FROM clean_cdr
GROUP BY branch

Running query in 'mysql+pyodbc://telecom'

8 rows affected.

branch  COUNT(*)
0      None    219390
1       中江县   8007501
2       什邡市   7339788
3       广汉市   9933897
4    德阳市市辖区  18487498
5  德阳市未知营业区      9668
6       绵竹市   6531970
7       罗江县   2591313

- distribution of the phone call's end time

In [13]:
%%sql

SELECT
    time,
    COUNT(*)
FROM clean_cdr
GROUP BY time
ORDER BY time

Running query in 'mysql+pyodbc://telecom'

24 rows affected.

time  COUNT(*)
0      0    290954
1      1    142586
2      2     84071
3      3     59412
4      4     58319
5      5    108623
6      6    514265
7      7   1527213
8      8   3256009
9      9   4338714
10    10   4475823
11    11   4469264
12    12   3319817
13    13   3234183
14    14   2981586
15    15   2781609
16    16   2892539
17    17   3513503
18    18   4138644
19    19   3501394
20    20   3260863
21    21   2332974
22    22   1276684
23    23    561976

- distribution of phone call's day of week

In [14]:
%%sql

SELECT
    day_of_week,
    COUNT(*)
FROM clean_cdr
GROUP BY day_of_week
ORDER BY day_of_week

Running query in 'mysql+pyodbc://telecom'

7 rows affected.

day_of_week  COUNT(*)
0            1   6341472
1            2   7043188
2            3   7005202
3            4   6832250
4            5   8735949
5            6   8834986
6            7   8327978

In [25]:
%%sql 
SELECT  
    acct_item_type_code, COUNT(*)
FROM clean_cdr
GROUP BY acct_item_type_code

Running query in 'mysql+pyodbc://telecom'

9 rows affected.

acct_item_type_code  COUNT(*)
0              030105      6088
1              160101  22617798
2              160201  23017227
3              160203         5
4              160204   3205179
5              160501   2099771
6              160603   1248301
7              160604    926654
8              170212         2

In [24]:
%sql SELECT DISTINCT etl_type_id FROM clean_cdr

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

etl_type_id
0          21

- the reamined area codes no matter the calling or called are all 0838 which indicates that all phone calls are within deyang

In [18]:
%%sql

SELECT
    client_nbr, calling_area_code, COUNT(*)
FROM clean_cdr
WHERE client_nbr = calling_nbr
GROUP BY client_nbr, calling_area_code

Running query in 'mysql+pyodbc://telecom'

417043 rows affected.

client_nbr calling_area_code  COUNT(*)
0        27h4lxfq              0838       243
1        27h4lxla              0838         4
2        27h4lxo2              0838        52
3        27h4lxqu              0838        34
4        27h4lxtm              0838        40
...           ...               ...       ...
417038   njzp3nf3              0838         4
417039   njzp3nkn              0838       115
417040   njzp3nnf              0838        14
417041   njzp3nsz              0838        41
417042   njzp3nvr              0838        14

[417043 rows x 3 columns]

In [20]:
%%sql

SELECT DISTINCT
    calling_area_code
FROM(
    SELECT
        client_nbr, calling_area_code, COUNT(*)
    FROM clean_cdr
    WHERE client_nbr = calling_nbr
    GROUP BY client_nbr, calling_area_code
) AS subquery

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

calling_area_code
0              0838

In [21]:
%%sql

SELECT
    client_nbr, called_area_code, COUNT(*)
FROM clean_cdr
WHERE client_nbr = called_nbr
GROUP BY client_nbr, called_area_code

Running query in 'mysql+pyodbc://telecom'

409098 rows affected.

client_nbr called_area_code  COUNT(*)
0        27h4lxfq             0838        75
1        27h4lxla             0838         9
2        27h4lxo2             0838       118
3        27h4lxqu             0838        41
4        27h4lxtm             0838        26
...           ...              ...       ...
409093   njzp3nf3             0838         9
409094   njzp3nkn             0838       138
409095   njzp3nnf             0838         3
409096   njzp3nsz             0838        73
409097   njzp3nvr             0838         5

[409098 rows x 3 columns]

In [22]:
%%sql

SELECT DISTINCT
    called_area_code
FROM(
    SELECT
        client_nbr, called_area_code, COUNT(*)
    FROM clean_cdr
    WHERE client_nbr = called_nbr
    GROUP BY client_nbr, called_area_code
) AS subquery

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

called_area_code
0             0838

## Check the meaning of ACC_NBR
- can't find the construction rule for client nubmer because when both the calling number and called number are client, client number can be calling number or called number

In [2]:
%%sql

SELECT COUNT(*)
FROM clean_cdr
WHERE 
    calling_nbr IN (SELECT MSISDN FROM tb_asz_cdma_0838_201308) AND
    called_nbr IN (SELECT MSISDN FROM tb_asz_cdma_0838_201308)

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0  13244425

In [4]:
%%sql

SELECT COUNT(*)
FROM clean_cdr
WHERE 
    calling_nbr IN (SELECT MSISDN FROM tb_asz_cdma_0838_201308) AND
    called_nbr IN (SELECT MSISDN FROM tb_asz_cdma_0838_201308) AND
    client_nbr = calling_nbr

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0   6597536

In [5]:
%%sql

SELECT COUNT(*)
FROM clean_cdr
WHERE 
    calling_nbr IN (SELECT MSISDN FROM tb_asz_cdma_0838_201308) AND
    called_nbr IN (SELECT MSISDN FROM tb_asz_cdma_0838_201308) AND
    client_nbr = called_nbr

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0   6647008

In [16]:
%%sql

SELECT calling_nbr, called_nbr, COUNT(*) AS count
FROM clean_cdr
GROUP BY calling_nbr, called_nbr
ORDER BY count

Running query in 'mysql+pyodbc://telecom'

12005982 rows affected.

calling_nbr called_nbr  count
0           njz85kbn   fm9461u4      1
1           njz85kbn   gcxrjn9k      1
2           njz85kbn   iagfexz5      1
3           njz85kbn   k82ejfxq      1
4           njz85kbn   m1nw9a1j      1
...              ...        ...    ...
12005977    ec23fg9f   7wkrfg9c   1204
12005978    iq4xjsfp   itx04xbt   1226
12005979    a9gbfna5   5z4rfna3   1268
12005980    5z4rfna3   a9gbfna5   1746
12005981    h4g48lys   6r789eig   1822

[12005982 rows x 3 columns]

In [19]:
%%sql

SELECT
    MAX(duration) AS column_value,
    decile
FROM (
    SELECT
        duration,
        NTILE(10) OVER (ORDER BY duration) AS decile
    FROM clean_cdr
) AS subquery
GROUP BY 
    decile

Running query in 'mysql+pyodbc://telecom'

10 rows affected.

column_value  decile
0            13       1
1            19       2
2            25       3
3            32       4
4            41       5
5            53       6
6            71       7
7           104       8
8           186       9
9          7201      10

In [82]:
%config SqlMagic.named_parameters=True

In [35]:
%sql DROP TABLE clean_cdr

Running query in 'mysql+pyodbc://telecom'

Empty DataFrame
Columns: []
Index: []

In [36]:
%%sql

CREATE TABLE clean_cdr(
    SELECT
        SERV_ID                                                                     AS id,
        ACC_NBR                                                                     AS client_nbr, 
        CONVERT(SUBSTRING_INDEX(SUBSTRING_INDEX(END_TIME,' ',-1),':',1),UNSIGNED)   AS time,
        DAYOFWEEK(SUBSTRING_INDEX(END_TIME,' ',1))                                  AS day_of_week,
        CALLING_NBR                                                                 AS calling_nbr,
        DURATION                                                                    AS duration,
        CALLED_NBR                                                                  AS called_nbr, 
        ACCT_ITEM_TYPE_CODE                                                         AS acct_item_type_code,
        CELL_ID                                                                     AS cell_id
    FROM serv_acct_item_0838_201308
    WHERE 
        ACCT_ITEM_TYPE_CODE != -1                                                  AND
        (START_TIME LIKE "____/%/% %:__:__" AND END_TIME LIKE "____/%/% %:__:__")  AND
        (CALLING_NBR NOT LIKE 'NotMobile_%' AND CALLED_NBR NOT LIKE 'NotMobile_%') AND
        (CALLING_AREA_COdE != '-1' AND CALLED_AREA_CODE != '-1')                   AND
        (LAC != '-1' AND LAC != '0')                                               AND
        (CELL_ID != '-1' AND CELL_ID != '0')                                       AND
        (calling_nbr in (SELECT MSISDN FROM tb_asz_cdma_0838_201308))              AND
        (called_nbr in (SELECT MSISDN FROM tb_asz_cdma_0838_201308))
)

Running query in 'mysql+pyodbc://telecom'

13244425 rows affected.

Empty DataFrame
Columns: []
Index: []

- there is no effect of including or excluding the cell_id exception on area code

In [37]:
%%sql

CREATE TABLE clean_cdr_area(
    SELECT
        SERV_ID                                                                     AS id,
        ACC_NBR                                                                     AS client_nbr, 
        CONVERT(SUBSTRING_INDEX(SUBSTRING_INDEX(END_TIME,' ',-1),':',1),UNSIGNED)   AS time,
        DAYOFWEEK(SUBSTRING_INDEX(END_TIME,' ',1))                                  AS day_of_week,
        CALLING_NBR                                                                 AS calling_nbr,
        CALLING_AREA_CODE                                                           AS calling_area_code,
        DURATION                                                                    AS duration,
        CALLED_AREA_CODE                                                            AS called_area_code,
        CALLED_NBR                                                                  AS called_nbr, 
        ACCT_ITEM_TYPE_CODE                                                         AS acct_item_type_code,
        CELL_ID                                                                     AS cell_id
    FROM serv_acct_item_0838_201308
    WHERE 
        ACCT_ITEM_TYPE_CODE != -1                                                  AND
        (START_TIME LIKE "____/%/% %:__:__" AND END_TIME LIKE "____/%/% %:__:__")  AND
        (CALLING_NBR NOT LIKE 'NotMobile_%' AND CALLED_NBR NOT LIKE 'NotMobile_%') AND
        (CALLING_AREA_COdE != '-1' AND CALLED_AREA_CODE != '-1')                   AND
        (LAC != '-1' AND LAC != '0')                                               AND
        (calling_nbr in (SELECT MSISDN FROM tb_asz_cdma_0838_201308))              AND
        (called_nbr in (SELECT MSISDN FROM tb_asz_cdma_0838_201308))
)

Running query in 'mysql+pyodbc://telecom'

13244425 rows affected.

There's a new jupysql version available (0.10.0), you're running 0.9.0. To upgrade: pip install jupysql --upgrade


Empty DataFrame
Columns: []
Index: []

In [52]:
%sql DROP TABLE clean_cdr_

Running query in 'mysql+pyodbc://telecom'

Empty DataFrame
Columns: []
Index: []

In [38]:
%sql SELECT DISTINCT calling_area_code FROM clean_cdr_area

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

calling_area_code
0              0838

In [39]:
%sql SELECT DISTINCT called_area_code FROM clean_cdr_area

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

called_area_code
0             0838

- 0838: 四川德陽市
- 028: 成都
- 0816: 四川綿陽市
- 023: 重慶
- 010: 北京
- 021: 上海
- 0817: 四川南充市
- 0755: 廣東深圳市
- 0825: 四川遂寧市
- 0833: 四川樂山市
- 020: 廣州
- 029: 陝西西安、咸陽市
- 028: 成都、眉山、資陽(028M, 028Z)
- 0839: 四川廣元
- 0834: 四川涼山市
- 0832: 四川內江市(0832a)
- 0830: 四川瀘州市
- 0871: 雲南昆明市
- 0813: 四川自貢市

- 0831: 四川宜賓市
- 0835: 四川雅安市

In [44]:
%%sql

SELECT
    CALLING_AREA_CODE, COUNT(*)
FROM serv_acct_item_0838_201308
WHERE LAC = '-1'
GROUP BY CALLING_AREA_CODE
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

393 rows affected.

CALLING_AREA_CODE  COUNT(*)
0                  -1  65121659
1                0838  40712679
2                 028   1257144
3                0816    251994
4                 023    115762
..                ...       ...
388              0027         1
389              0041         1
390              0046         1
391             00855         1
392             00968         1

[393 rows x 2 columns]

In [45]:
%%sql

SELECT
    CALLING_AREA_CODE, COUNT(*)
FROM serv_acct_item_0838_201308
WHERE LAC = '0'
GROUP BY CALLING_AREA_CODE
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

395 rows affected.

CALLING_AREA_CODE  COUNT(*)
0                0838   1666204
1                 028     62928
2                 023     49165
3                0991     20064
4                 010     18985
..                ...       ...
390             00353         1
391              0041         1
392             00679         1
393             00855         1
394              0090         1

[395 rows x 2 columns]

In [49]:
%%sql

SELECT
    CALLING_AREA_CODE, COUNT(*)
FROM serv_acct_item_0838_201308
WHERE LAC = '0' AND CALLING_AREA_CODE LIKE '00%'
GROUP BY CALLING_AREA_CODE
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

63 rows affected.

CALLING_AREA_CODE  COUNT(*)
0              00852       195
1                007       104
2              00886        65
3                001        52
4               0066        35
..               ...       ...
58            001418         1
59             00855         1
60             00352         1
61             00250         1
62             00254         1

[63 rows x 2 columns]

In [56]:
%%sql

SELECT
    COUNT(*)
FROM serv_acct_item_0838_201308
WHERE LAC = '0' AND CALLING_AREA_CODE = '0838' 

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0   1666204

In [59]:
%%sql

SELECT
    COUNT(*)
FROM serv_acct_item_0838_201308
WHERE LAC = '0' AND CALLING_AREA_CODE = '0838' AND CALLED_AREA_CODE = '0838'

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0    900108

In [60]:
%%sql

SELECT
    COUNT(*)
FROM serv_acct_item_0838_201308
WHERE 
    LAC != 0 AND LAC != -1 AND 
    CALLING_AREA_CODE != '0838' AND CALLED_AREA_CODE != '0838' AND 
    CALLING_AREA_CODE != -1 AND CALLED_AREA_CODE != '-1'

Running query in 'mysql+pyodbc://telecom'

1 rows affected.

COUNT(*)
0         0

In [47]:
%%sql

SELECT
    CALLED_AREA_CODE, COUNT(*)
FROM serv_acct_item_0838_201308
WHERE LAC = '0'
GROUP BY CALLED_AREA_CODE
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

347 rows affected.

CALLED_AREA_CODE  COUNT(*)
0               0838   1570579
1                028     74785
2                023     55833
3                010     21161
4               0991     20748
..               ...       ...
342             0033         1
343             0062         1
344             0065         1
345            00686         1
346            00998         1

[347 rows x 2 columns]

In [50]:
%%sql

SELECT
    CALLED_AREA_CODE, COUNT(*)
FROM serv_acct_item_0838_201308
WHERE LAC = '0' AND CALLED_AREA_CODE LIKE '00%'
GROUP BY CALLED_AREA_CODE
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

18 rows affected.

CALLED_AREA_CODE  COUNT(*)
0             00852        24
1               001        15
2              0081         9
3              0061         8
4             00886         7
5             00853         6
6              0082         4
7            001514         4
8              0049         3
9               007         2
10             0084         2
11            00255         2
12            00998         1
13            00686         1
14           001613         1
15             0033         1
16             0062         1
17             0065         1

In [6]:
%sql DROP TABLE clean_cdr

Running query in 'mysql+pyodbc://telecom'

Empty DataFrame
Columns: []
Index: []

In [4]:
%config SqlMagic.named_parameters=True

In [7]:
%%sql

CREATE TABLE clean_cdr(
    SELECT
        CONVERT(SUBSTRING_INDEX(SUBSTRING_INDEX(END_TIME,' ',-1),':',1),UNSIGNED)   AS time,
        DAYOFWEEK(SUBSTRING_INDEX(END_TIME,' ',1))                                  AS day_of_week,
        CALLING_NBR                                                                 AS calling_nbr,
        CALLING_AREA_CODE                                                           AS calling_area_code, 
        DURATION                                                                    AS duration,
        CALLED_AREA_CODE                                                            AS called_area_code, 
        CALLED_NBR                                                                  AS called_nbr, 
        CELL_ID                                                                     AS cell_id
    FROM serv_acct_item_0838_201308
    WHERE 
        ACCT_ITEM_TYPE_CODE != -1                                                  AND
        (START_TIME LIKE "____/%/% %:__:__" AND END_TIME LIKE "____/%/% %:__:__")  AND
        (CALLING_NBR NOT LIKE 'NotMobile_%' AND CALLED_NBR NOT LIKE 'NotMobile_%') AND
        (CALLING_AREA_COdE != '-1' AND CALLED_AREA_CODE != '-1')                   AND
        (CELL_ID != '-1' AND CELL_ID != '0')                                       AND
        (CALLING_NBR in (SELECT MSISDN FROM tb_asz_cdma_0838_201308) OR CALLED_NBR in (SELECT MSISDN FROM tb_asz_cdma_0838_201308)) AND
        (CALLING_AREA_CODE = '0838' OR CALLED_AREA_CODE = '0838')
)

Running query in 'mysql+pyodbc://telecom'

52978357 rows affected.

Empty DataFrame
Columns: []
Index: []

In [8]:
%%sql 

SELECT calling_area_code, COUNT(*)
FROM clean_cdr
GROUP BY calling_area_code
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

358 rows affected.

calling_area_code  COUNT(*)
0                0838  48903288
1                 028   1859392
2                0816    397343
3                 023    114454
4                0825     61418
..                ...       ...
353              0093         2
354              0036         1
355              0040         1
356             00880         1
357              0092         1

[358 rows x 2 columns]

In [9]:
%%sql

SELECT called_area_code, COUNT(*)
FROM clean_cdr
GROUP BY called_area_code
ORDER BY COUNT(*) DESC

Running query in 'mysql+pyodbc://telecom'

331 rows affected.

called_area_code  COUNT(*)
0               0838  48485966
1                028   2266889
2               0816    465680
3                023    137094
4               0817     80407
..               ...       ...
326             0457        97
327             0464        62
328             0049         3
329             0062         2
330             0283         1

[331 rows x 2 columns]